lib/spack/llnl/util/filesystem.py

```diff
@@ -924,6 +924,71 @@ def traverse_tree(source_root, dest_root, rel_path='', **kwargs):
         yield (source_path, dest_path)
 
 
+def visit_directory_tree(root, visitor, rel_path='', depth=0):
+    """
+    Recurses the directory root depth-first through a visitor pattern
+
+    The visitor interface is as follows:
+    - handle_file(root, rel_path, depth) -> bool
+        if False is returned, the iteration over children in the current dir is stopped
+    - before_enter_dir(root, rel_path, depth) -> bool
+        if True is returned, descends into this directory
+    - before_enter_symlinked_dir(root, rel_path, depth) -> bool
+        if True is returne,d descends into this directory
+    - after_enter_dir(root, rel_path, depth) -> void
+        only called when before_enter_dir returns True
+    - after_enter_symlinked_dir(root, rel_path, depth) -> void
+        only called when before_enter_symlinked_dir returns True
```

Comment:
I would rename all of these `enter` -> `visit`. When I first read this, I thought `after_enter_dir` meant it was run immediately after entering the directory and before recursing on its contents.

We should also consider whether an option to handle symlinks differently than other files would be good for the interface, even though we handle them the same in the view use-case